File: SecondYearCS/OOP/CivitasRB/lib/civitas_juego.rb
require_relative 'jugador.rb'
require_relative 'gestor_estados.rb'
require_relative 'mazo_sorpresa.rb'
require_relative 'CasillaDescanso.rb'
require_relative 'CasillaSorpresa.rb'
require_relative 'CasillaCalle.rb'
require_relative 'CasillaImpuesto.rb'
require_relative 'CasillaJuez.rb'
require_relative 'SorpresaPagarCobrar.rb'
require_relative 'SorpresaJugadorEspeculador.rb'
require_relative 'SorpresaIrCasilla.rb'
require_relative 'SorpresaPorCasaHotel.rb'
require_relative 'SorpresaPorJugador.rb'
require_relative 'SorpresaSalirCarcel.rb'
require_relative 'SorpresaIrCarcel.rb'
require_relative 'titulo_propiedad.rb'
require_relative 'diario.rb'
require_relative 'dado.rb'
require_relative 'tablero.rb'
require_relative 'operaciones_juego.rb'
module CivitasRB
  class CivitasJuego
   
    def initialize(args)
      @jugadores = Array.new 
      args.each do |n|
        @jugadores << Jugador.new(n)
      end
      @indiceJugadorActual = Dado.instance.quienEmpieza(@jugadores.length)
      @gestorEstados = Gestor_estados.new
      @estado = @gestorEstados.estado_inicial
      @mazo = MazoSorpresa.new
      @@tabSize = 6
      self.inicializaTablero
      self.inicializaMazoSorpresa
    end
    
   
    
    #Estoy mas perdido que estefania
    def inicializaTablero
     @tablero = Tablero.new(@@tabSize)
     i = 1
     loop do
       if(i==3 ||i == 13 ||i == 18)
         @tablero.anadeCasilla(CasillaSorpresa.new(@mazo,"Sorpresa"))
       elsif(i==6)
         #nothing
       elsif(i==8)
         @tablero.anadeCasilla(CasillaImpuesto.new( 0.15*1000,"Impuesto"))
       elsif(i==11)
         @tablero.anadeCasilla(CasillaDescanso.new("Parking"))
       elsif(i==16)
         @tablero.anadeJuez()
       elsif(i==20)
         break
       else
         @tablero.anadeCasilla(CasillaCalle.new(TituloPropiedad.new("Calle"+ i.to_s,25+i*25,0.5+i*0.1,40+i*5,40+i*20,60+i*5),"CasillaCalle"))
       end
       i+=1
     end
    end
    
    def inicializaMazoSorpresa
      @mazo.alMazo(SorpresaIrCasilla.new(@tablero,10,"IrCasilla"))
      @mazo.alMazo(SorpresaIrCasilla.new(@tablero,3,"IrCasilla"))
      @mazo.alMazo(SorpresaJugadorEspeculador.new("ConvertirAJugadorMafioso",300))
      @mazo.alMazo(SorpresaJugadorEspeculador.new("ConvertirAJugadorEntrepeneu",150))
      @mazo.alMazo(SorpresaIrCarcel.new(@tablero,"IrACarcelPringao"))
      @mazo.alMazo(SorpresaIrCarcel.new(@tablero,"IrACarcelPringao2"))
      @mazo.alMazo(SorpresaPagarCobrar.new(75,"PagarCobrar"))
      @mazo.alMazo(SorpresaSalirCarcel.new(@mazo,"PorChivato"))
      @mazo.alMazo(SorpresaSalirCarcel.new(@mazo,"ErrorPapeleo"))
      @mazo.alMazo(SorpresaPorCasaHotel.new(10,"APagarImpuesto"))
      @mazo.alMazo(SorpresaPorCasaHotel.new(10,"ErrorImpuesto"))
      @mazo.alMazo(SorpresaPorJugador.new(30,"HasGanadoAlPoker"))
      @mazo.alMazo(SorpresaPorJugador.new(-30,"HasPerdidoAlPoker"))
    end
    
    def contabilizarPasosPorSalida(jugadorActual)
      while @tablero.getPorSalida > 0
        jugadorActual.pasaPorSalida
      end
    end
    
    def pasarTurno()
      @indiceJugadorActual += 1
      if(@indiceJugadorActual >= @jugadores.length)
        @indiceJugadorActual = 0;
      end
    end
    
    #Ni zorra de lo que hay que hacer aqui, me he inventado
    #EL atributo estado y bueno ahi va 
    def siguientePasoCompletado(operacion)
      @estado = @gestorEstados.siguiente_estado(@jugadores[@indiceJugadorActual], @estado, operacion)
    end
    
    #Para implementar este hay que implementarlos en la 
    #clase jugador ( Verificar si funciona allí )
    def construirCasa(ip)
      @jugadores[@indiceJugadorActual].construirCasa(ip)
    end
    
    def construirHotel(ip)
      @jugadores[@indiceJugadorActual].construirHotel(ip)
    end
    
    def vender(ip)
      @jugadores[@indiceJugadorActual].vender(ip)
    end
    
    def hipotecar(ip)
      @jugadores[@indiceJugadorActual].hipotecar(ip)
    end

    def cancelarHipoteca(ip)
      @jugadores[@indiceJugadorActual].cancelarHipoteca(ip)
    end
    
    def salirCarcelPagando()
      @jugadores[@indiceJugadorActual].salirCarcelPagando()
    end
    
    def salirCarcelTirando()
      @jugadores[@indiceJugadorActual].salirCarcelTirando()
    end
    
    #Me he inventado el bucle for, igual tengo suerte
    def finalDelJuego()
      for player in @jugadores do
        if ( player.enBancarrota() )
          return true
        end
      end
      return false
    end
    
    def ranking
      return @jugadores.sort.reverse #Ni idea si funcionará o no, tengo fé
    end
    
    def avanzaJugador()
      jugadorActual = @jugadores.at(@indiceJugadorActual)
      posicionActual = jugadorActual.numCasillaActual
      tirada = Dado.instance.tirar
      posicionNueva = @tablero.nuevaPosicion(posicionActual, tirada)
      casilla = @tablero.getCasilla(posicionNueva)
      self.contabilizarPasosPorSalida(jugadorActual)
      jugadorActual.moverACasilla(posicionNueva)
      casilla.recibeJugador(@indiceJugadorActual,@jugadores)
      self.contabilizarPasosPorSalida(jugadorActual)
    end
    
    def siguientePaso()
      jugadorActual = @jugadores.at(@indiceJugadorActual)
      operacion = @gestorEstados.operaciones_permitidas(jugadorActual, @estado)
      if(operacion==Operaciones_juego::PASAR_TURNO)
        self.pasarTurno
        self.siguientePasoCompletado(operacion)
      end
      if(operacion==Operaciones_juego::AVANZAR)
        self.avanzaJugador
        self.siguientePasoCompletado(operacion)
      end
      return operacion
    end
    
    def comprar()
      jugadorActual = @jugadores.at(@indiceJugadorActual)
      numCasillaActual = jugadorActual.numCasillaActual
      casilla = @tablero.getCasilla(numCasillaActual)
      titulo = casilla.tituloPropiedad
      result = jugadorActual.comprar(titulo)
      return result
    end

    def getJugador
      return @jugadores.at(@indiceJugadorActual)
    end

    def getCasillaActual
      numCasillaActual =  @jugadores.at(@indiceJugadorActual).numCasillaActual
      return @tablero.getCasilla(numCasillaActual)
    end

  end
end

File: SecondYearCS/OOP/Civitas/src/civitas/CivitasJuego.java
package civitas;
import GUI.Dado;
import java.util.ArrayList;
import java.util.Collections;
public class CivitasJuego {
    
    private ArrayList<Jugador> jugadores = new ArrayList<>();
    private int indiceJugadorActual;
    private GestorEstados gestorEstados;
    private EstadosJuego estado;
    private Tablero tablero;
    private int carcel=5;
    private MazoSorpresa mazo;
    
    public CivitasJuego(ArrayList<String> nombres){
        for(int i=0;i<nombres.size();i++){
            Jugador nuevo = new Jugador(nombres.get(i));
            jugadores.add(nuevo);
        }
        gestorEstados = new GestorEstados();
        this.estado = gestorEstados.estadoInicial();
        indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
        mazo = new MazoSorpresa(true);
        this.inicializaTablero(this.mazo);
        this.inicializaMazoSorpresa(this.tablero);
    }
    
    private void inicializaTablero(MazoSorpresa mazo){
        this.tablero = new Tablero(carcel);
        for(int i=1;i<20;i++){
            if(i==3 || i == 12 | i == 17)
                this.tablero.añadeCasilla(new CasillaSorpresa(this.mazo,"Sorpresa"));
            else if(i==6)
                continue;
            else if(i==7)
                this.tablero.añadeCasilla(new CasillaImpuesto(0.15f*1000f,"Impuesto"));
            else if(i==10)
                this.tablero.añadeCasilla(new CasillaDescanso("Parking"));
            else if(i==15)
                this.tablero.añadeJuez();
            else 
                this.tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle"+Integer.toString(i),25 + i*25,0.5f + i*0.1f,40 + i*5,40 + i*20,60 + i*5),"Calle")); //Values?¿
        }
       

    }
    
    /**
     * @param tablero 
     * EL profesor inicializa sus sorpresas con x lineas de una simple
     * llamada a mazo.AlMazo, de manera que vamos a ignorar nuestro bucle y 
     * realizarlo de esta manera tambien.
     */
    private void inicializaMazoSorpresa(Tablero tablero){
       mazo.alMazo(new SorpresaIrCasilla(tablero,10,"IrACasilla"));
       mazo.alMazo(new SorpresaIrCasilla(tablero,3,"IrACasilla"));
       mazo.alMazo(new SorpresaJugadorEspeculador("ConvertirAJugadorMafioso",300f));
       mazo.alMazo(new SorpresaJugadorEspeculador("ConvertirAJugadorEntrepeneur",150f));
       mazo.alMazo(new SorpresaIrCarcel(tablero,"IrACarcelPringao"));
       mazo.alMazo(new SorpresaIrCarcel(tablero,"IrACarcelPringao2"));
       mazo.alMazo(new SorpresaPagarCobrar(75f,"PagarCobrar"));
       mazo.alMazo(new SorpresaSalirCarcel(this.mazo,"PorChivato"));
       mazo.alMazo(new SorpresaSalirCarcel(this.mazo,"ErrorPapeleo"));
       mazo.alMazo(new SorpresaPorCasaHotel(10f,"APagarImpuesto"));
       mazo.alMazo(new SorpresaPorCasaHotel(-10f,"ErrorImpuesto"));
       mazo.alMazo(new SorpresaPorJugador(30f,"HasGanadoAlPoker"));
       mazo.alMazo(new SorpresaPorJugador(-30f,"HasPerdidoAlPoker"));
    }
    
    public void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(this.tablero.getPorSalida()>0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    private void pasarTurno(){
        this.indiceJugadorActual++;
        if(this.indiceJugadorActual >= this.jugadores.size())
            indiceJugadorActual = 0;
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        this.estado = this.gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), this.estado, operacion);
  
    }
    
    public boolean construirCasa(int ip){
        return this.jugadores.get(this.indiceJugadorActual).construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
        return this.jugadores.get(this.indiceJugadorActual).construirHotel(ip);
    }
    
    public boolean vender(int ip){
        return this.jugadores.get(this.indiceJugadorActual).vender(ip);
    }
    
    public boolean hipotecar(int ip){
        return this.jugadores.get(this.indiceJugadorActual).hipotecar(ip);
    }
    
    public boolean cancelarHipoteca(int ip){
        return this.jugadores.get(this.indiceJugadorActual).cancelarHipoteca(ip);
    }
    
    public boolean salirCarcelPagando(){
        return this.jugadores.get(this.indiceJugadorActual).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return this.jugadores.get(this.indiceJugadorActual).salirCarcelTirando();
    }
    
    public boolean finalDelJuego(){
        boolean result = false;
        for(Jugador jugador: this.jugadores){
            if(jugador.enBancarrota())
                result = true;
        }
        return result;
    }
    
    public ArrayList<Jugador> ranking(){
        Collections.sort(this.jugadores);
        Collections.reverse(this.jugadores);
        return this.jugadores;
    }
    
    private void avanzaJugador() throws Exception{
        Jugador jugadorActual = this.jugadores.get(indiceJugadorActual);
        int posicionActual = jugadorActual.getNumCasillaActual();
        int tirada = Dado.getInstance().tirar();
        int posicionNueva = this.tablero.nuevaPosicion(posicionActual,tirada);
        Casilla casilla = this.tablero.getCasilla(posicionNueva);
        this.contabilizarPasosPorSalida(jugadorActual);
        jugadorActual.moverACasilla(posicionNueva);
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        this.contabilizarPasosPorSalida(jugadorActual);
    }
    
    public OperacionesJuego siguientePaso() throws Exception{
        Jugador jugadorActual = this.jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
        if(operacion==OperacionesJuego.PASAR_TURNO){
            this.pasarTurno();
            this.siguientePasoCompletado(operacion);
        }else if (operacion==OperacionesJuego.AVANZAR){
            this.avanzaJugador();
            this.siguientePasoCompletado(operacion);
        }
        return operacion;
    }
    
    public boolean comprar(){
        boolean result = false;
        Jugador jugadorActual = this.jugadores.get(indiceJugadorActual);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        Casilla casilla = this.tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo = ((CasillaCalle)casilla).getTituloPropiedad();
        result = jugadorActual.comprar(titulo);
        return result;
    }
    
    public Jugador getJugador(){
        return this.jugadores.get(this.indiceJugadorActual);
    }
    
    public Casilla getCasillaActual(){
        Jugador jugadorActual = this.jugadores.get(this.indiceJugadorActual);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        return this.tablero.getCasilla(numCasillaActual);
    }
    
    public int numCasillaActual(){
        return this.jugadores.get(this.indiceJugadorActual).getNumCasillaActual();
    }
    
}
